### PR TITLE
Load should base on the accurate loaded number

### PR DESCRIPTION
**Pull request by smellthemoon**
Signed-off-by: lixinguo <xinguo.li@zilliz.com>

### CODE REVIEW

File: internal/querycoordv2/observers/collection_observer.go [hunk 1]
```diff
@@ -167,6 +176,7 @@ func (ob *CollectionObserver) observeCollectionLoadStatus(collection *meta.Colle
 	if targetNum == 0 {
 		log.Info("No segment/channel in target need to be loaded!")
 		updated.LoadPercentage = 100
+		delete(ob.collectionLoadedMap, collection.CollectionID)
```
Comment (yah01) [inline internal/querycoordv2/observers/collection_observer.go:179]: Move this to line 143

File: internal/querycoordv2/observers/collection_observer.go [hunk 2]
```diff
@@ -154,6 +160,9 @@ func (ob *CollectionObserver) observeLoadStatus() {
 func (ob *CollectionObserver) observeCollectionLoadStatus(collection *meta.Collection) {
 	log := log.With(zap.Int64("collectionID", collection.GetCollectionID()))
 
+	if _, ok := ob.collectionLoadedMap[collection.CollectionID]; !ok {
+		ob.collectionLoadedMap[collection.CollectionID] = 0
+	}
```
Comment (yah01) [inline internal/querycoordv2/observers/collection_observer.go:165]: The initialization is not needed

File: internal/querycoordv2/observers/collection_observer.go [hunk 3]
```diff
@@ -34,9 +34,11 @@ import (
 type CollectionObserver struct {
 	stopCh chan struct{}
 
-	dist      *meta.DistributionManager
-	meta      *meta.Meta
-	targetMgr *meta.TargetManager
+	dist                *meta.DistributionManager
+	meta                *meta.Meta
+	targetMgr           *meta.TargetManager
+	collectionLoadedMap map[int64]int
```
Comment (yah01) [inline internal/querycoordv2/observers/collection_observer.go:40]: `collectionLoadedMap` -> `collectionLoadedCount`, The type has implied it's a map

File: internal/querycoordv2/observers/collection_observer.go [hunk 4]
```diff
@@ -45,12 +47,16 @@ func NewCollectionObserver(
 	dist *meta.DistributionManager,
 	meta *meta.Meta,
 	targetMgr *meta.TargetManager,
+	collectionLoadedMap map[int64]int,
+	partitionLoadedMap map[int64]int,
```
Comment (yah01) [inline internal/querycoordv2/observers/collection_observer.go:51]: these new fields shouldn't be added to the constructor parameters, they are internal states, just initialize them at line 58

**Comment by yah01** [comment]
Also need a unit test to check whether the `collection.UpdatedAt` is updated while only segments less than 1% were loaded

**Comment by yah01** [comment]
related #20775, with the two PRs, load will make progress as any segment loaded

File: internal/querycoordv2/observers/collection_observer.go (hunk 4 above)
Comment (smellthemoon) [reply internal/querycoordv2/observers/collection_observer.go:51]: ok

File: internal/querycoordv2/observers/collection_observer.go [hunk 5]
```diff
@@ -195,10 +201,10 @@ func (ob *CollectionObserver) observeCollectionLoadStatus(collection *meta.Colle
 		updated.LoadPercentage = int32(loadedCount * 100 / targetNum)
 	}
 
-	if updated.LoadPercentage <= collection.LoadPercentage {
+	if loadedCount <= ob.collectionLoadedCount[collection.GetCollectionID()] && updated.LoadPercentage != 100 {
```
Comment (yah01) [inline internal/querycoordv2/observers/collection_observer.go:204]: `updated.LoadPercentage != 100` is not needed

File: internal/querycoordv2/observers/collection_observer.go [hunk 6]
```diff
@@ -260,10 +266,10 @@ func (ob *CollectionObserver) observePartitionLoadStatus(partition *meta.Partiti
 
 	}
 
-	if updated.LoadPercentage <= partition.LoadPercentage {
+	if loadedCount <= ob.partitionLoadedCount[partition.GetPartitionID()] && updated.LoadPercentage != 100 {
```
Comment (yah01) [inline internal/querycoordv2/observers/collection_observer.go:269]: same as line 204

File: internal/querycoordv2/observers/collection_observer_test.go [hunk 7]
```diff
@@ -232,18 +256,23 @@ func (suite *CollectionObserverSuite) isCollectionTimeout(collection int64) bool
...
 	segments := suite.targetMgr.GetHistoricalSegmentsByCollection(collection, meta.CurrentTarget)
-
 	return !(exist ||
 		len(replicas) > 0 ||
 		len(channels) > 0 ||
 		len(segments) > 0)
 }
 
+func (suite *CollectionObserverSuite) isCollectionLoadedContinue(collection int64, beforeTime time.Time) bool {
+	fmt.Println(suite.meta.GetCollection(collection).UpdatedAt, beforeTime, "aaaaaaaaa")
```
Comment (yah01) [inline internal/querycoordv2/observers/collection_observer_test.go:266]: remove this debug line

File: internal/querycoordv2/observers/collection_observer_test.go [hunk 8]
```diff
@@ -232,18 +256,23 @@ func (suite *CollectionObserverSuite) isCollectionTimeout(collection int64) bool
...
+	return suite.meta.GetCollection(collection).UpdatedAt != beforeTime
```
Comment (yah01) [inline internal/querycoordv2/observers/collection_observer_test.go:267]: suite.meta.GetCollection(collection).UpdatedAt.After(beforeTime) should be better

File: internal/querycoordv2/observers/collection_observer_test.go [hunk 9]
```diff
@@ -304,3 +333,17 @@ func (suite *CollectionObserverSuite) load(collection int64) {
 func TestCollectionObserver(t *testing.T) {
 	suite.Run(t, new(CollectionObserverSuite))
 }
+
+func genSegmentsInfo(count int, start int, collID int64, partitionID int64, insertChannel string) []*datapb.SegmentInfo {
+	var ret []*datapb.SegmentInfo
```
Comment (yah01) [inline internal/querycoordv2/observers/collection_observer_test.go:338]: `ret := make([]*datapb.SegmentInfo, 0, count)` can avoid many times coping

**Pull request merged**